src/amo/sagas/collections.js

```diff
@@ -166,43 +168,70 @@ export function* addAddonToCollection({
   }
 }
 
-export function* updateCollection({
+export function* modifyCollection({
+  type,
   payload: {
-    errorHandlerId,
     collectionSlug,
     defaultLocale,
     description,
+    errorHandlerId,
     formOverlayId,
     name,
     slug,
     user,
   },
-}: UpdateCollectionAction): Generator<any, any, any> {
+}: {
+  type: typeof CREATE_COLLECTION | typeof UPDATE_COLLECTION,
+  payload: {
+    ...RequiredModifyCollectionParams,
+    ...OptionalModifyCollectionParams,
+    collectionSlug?: string,
+    name?: ?LocalizedString,
+    slug?: ?string,
+  }
+}): Generator<any, any, any> {
+  const creating = type === CREATE_COLLECTION;
   const errorHandler = createErrorHandler(errorHandlerId);
   yield put(errorHandler.createClearingAction());
 
   try {
     yield put(beginFormOverlaySubmit(formOverlayId));
 
     const state = yield select(getState);
-    const params: UpdateCollectionParams = {
-      api: state.api,
-      collectionSlug,
-      defaultLocale,
-      description,
-      name,
-      slug,
-      user,
-    };
-    yield call(api.updateCollection, params);
 
-    const slugWasEdited = slug && slug !== collectionSlug;
-    const effectiveSlug = slug || collectionSlug;
+    if (creating) {
+      const apiParams = {
```

Comment:
That's how I originally had it, but I couldn't seem to get Flow happy with that in place and the other type definitions I had. When I removed it Flow didn't complain, so I assumed that meant it wasn't necessary. I'll try adding it back in and see if I can work through the Flow errors. I may post them here for some help.

Reply:
Ok, the problem was because in the type for the function parameters, I need to make `name`, `slug` and `collectionSlug` all optional, because some are optional for each of the two actions. That makes the types `CreateCollectionParams` and `UpdateCollectionParams` incompatible. I was able to solve this by adding some calls to `invariant` to ensure that they were not empty when they shouldn't be, and that makes Flow happy. I suppose there's nothing wrong with this, and is in fact appropriate, although I know the the only code we have that calls this also makes sure those properties are populated when needed, so the extra `invariant`s seem unnecessary. I suppose there's no guarantee that there won't be a future caller who doesn't ensure that, which is kind of the point of Flow in the first place.

So I guess this all makes sense and I'm happy with it, and I learned something too. 😄